Cesium3DTilesSelection/src/TilesetLoadTilesetDotJson.cpp

```diff
@@ -430,24 +444,156 @@ void Tileset::LoadTilesetDotJson::Private::workerThreadLoadTerrainTile(
+                return asyncSystem.createResolvedFuture();
+              }
+
+              if (pResponse->statusCode() != 0 &&
+                  (pResponse->statusCode() < 200 ||
+                   pResponse->statusCode() >= 300)) {
+                SPDLOG_LOGGER_ERROR(
+                    pLogger,
+                    "Received status code {} for parent layer.json {}",
+                    pResponse->statusCode(),
+                    pRequest->url());
+                return asyncSystem.createResolvedFuture();
+              }
+
+              const gsl::span<const std::byte> data = pResponse->data();
+              rapidjson::Document parentLayerJson;
+              parentLayerJson.Parse(
+                  reinterpret_cast<const char*>(data.data()),
+                  data.size());
+
+              if (parentLayerJson.HasParseError()) {
+                SPDLOG_LOGGER_ERROR(
+                    pLogger,
+                    "Error when parsing layer.json, error code {} at byte "
+                    "offset ",
+                    "{}",
+                    parentLayerJson.GetParseError(),
+                    parentLayerJson.GetErrorOffset());
+                return asyncSystem.createResolvedFuture();
+              }
+
+              context.pUnderlyingContext = std::make_unique<TileContext>();
+              context.pUnderlyingContext->baseUrl = pRequest->url();
+              context.pUnderlyingContext->pTileset = context.pTileset;
+              context.pUnderlyingContext->requestHeaders =
+                  context.requestHeaders;
+              if (context.pTopContext) {
+                context.pUnderlyingContext->pTopContext = context.pTopContext;
+              } else {
+                context.pUnderlyingContext->pTopContext = &context;
+              }
+
+              return Private::workerThreadLoadTileContext(
+                  parentLayerJson,
+                  *context.pUnderlyingContext,
+                  asyncSystem,
+                  pLogger,
+                  useWaterMask);
+            });
+  }
+  return asyncSystem.createResolvedFuture();
+}
+
+Future<void> Tileset::LoadTilesetDotJson::Private::workerThreadLoadTerrainTile(
+    Tile& tile,
+    const rapidjson::Document& layerJson,
+    TileContext& context,
+    const CesiumAsync::AsyncSystem& asyncSystem,
+    const std::shared_ptr<spdlog::logger>& pLogger,
+    bool useWaterMask) {
+  context.requestHeaders.push_back(std::make_pair(
+      "Accept",
+      "application/vnd.quantized-mesh,application/octet-stream;q=0.9,*/"
+      "*;q=0.01"));
+  return Private::workerThreadLoadTileContext(
+             layerJson,
+             context,
+             asyncSystem,
+             pLogger,
+             useWaterMask)
+      .thenImmediately([&tile, &context]() {
+        tile.setContext(&context);
+        tile.setBoundingVolume(
+            context.implicitContext->implicitRootBoundingVolume);
+        tile.setGeometricError(999999999.0);
```

Comment:
this isn't thread-safe, loadImmediately() will certainly run on worker thread, but tile must be modified at main-thread

The lambda also captures the context by reference, that means there is a good chance that two more threads are accessing the same context at the same time, which is not thread safe as well

Reply:
> Thanks @joseph-kaile! Most looks good to me. The only thing that concerns me is TilesetLoadTilesetDotJson isn't thread-safe at the moment as the worker threads are using Tile and TileContext by reference which are thread-safe themselves

Thanks for the suggestions.

I do think that the future code is thread safe, I don't think tile would be modified by any other thread because the root tile won't be set until these futures complete.

I dont think any of the tasks that are spawned actually run concurrently:

workerThreadLoadTerrainTile runs workerThreadLoadTileContext, and then waits for it to finish before continuing.

workerThreadLoadTileContext potentially spawns off workerThreadLoadTileContext when it has finished.

And there's nothing else out there that would be running at the same time.

Reply:
I may miss something, but workerThreadLoadTerrainTile is called from another worker thread at line 277. The original thread doesn't wait for workerThreadLoadTerrainTile to finish, but resolved right away, which mean the root tile is already ended up in the main thread while workerThreadLoadTileContext is running

Reply:
@baothientran you're right, I may have broken this when I made my changes in that other branch. Let me take a look.

Reply:
I see the problem now, I thought that I returned the future. Thanks.